tests/rhsm/test_rhsm_client.py

```diff
@@ -132,3 +132,48 @@ def test_create_image(requests_mocker, caplog):
     out = client.create_image("ami-123", "ami-rhel", "x86_64", "RHEL", "us-east-1")
     assert isinstance(out.exception(), ConnectionError)
     assert caplog.messages == ["Failed to process request to RHSM with exception "]
+
+
+def test_list_images(requests_mocker, caplog):
+    """
+    Test listing of all images from rhsm and checks requests sent and
+    data received while using pagination logic.
+    """
+    url = "https://example.com/v1/internal/cloud_access_providers/amazon/amis"
+    caplog.set_level(logging.DEBUG)
+
+    def create_response(amis_count, start):
+        return {
+            "status_code": 200,
+            "json": {
+                "pagination": {"count": amis_count},
+                "body": [
+                    {"amiID": f"ami-{i}"} for i in range(start, start + amis_count)
+                ],
+            },
+        }
+
+    responses = [
+        create_response(1000, 1),
+        create_response(1, 1001),
+        create_response(0, 1002),
+    ]
+
+    m_list_images = requests_mocker.register_uri("GET", url, responses)
+
+    client = RHSMClient(
+        "https://example.com", cert=("client.crt", "client.key"), max_retry_sleep=0.001
+    )
+
+    image_ids = client.list_image_ids()
+
+    # there should be 3 calls, last won't get any data, so we stop requesting another page.
+    assert m_list_images.call_count == 3
+    for req_history, offset in zip(m_list_images.request_history, [0, 1000, 2000]):
```

Comment:
I feel like the result here in this test demonstrates what I was saying earlier that it seems wrong to always increase the offset by 1000.

If I understand right what the test simulates:

- we request with offset=0 and get 1000 items back
- then we request with offset=1000 and get 1 more item back
- then we request with offset=2000 and get 0 items back

The last request doesn't seem to make sense to me, if the server only provided us 1001 items so far, isn't it odd to make the next request starting at item number 2000 ?